index.js

```
@@ -18,59 +22,122 @@ var tmpPrefix = "tessel-rust-compile"
-              }
+    // Create a stream for extracting the uploaded project archive
+    var extractStream = tar.extract(dirPath);
+    // Send the incoming project tarball into the extractor
+    req.pipe(extractStream);
+    // Once the extraction completes
+    extractStream.once('finish', function extractionComplete() {
+      // Read the contents of the extracted directory:
+      // -- temp_dir
+      // ---- deployed_project_filder
+      fs.readdir(dirPath, (readDirError, contents) => {
+
+        // If we were unable to read the directory
+        if (readDirError !== null) {
+          // Let the client know that we had an issue
+          return rejectRequest(responseCodes.HTTP_INTERNAL_SERVER_ERROR, readDirError);
+        }
+
+        // There should only be a single folder (the deployed folder)
+        if (contents.length !== 1) {
+          // If there zero or more than one, reject and notify client
+          return rejectRequest(responseCodes.HTTP_BAD_REQUEST,
+            'Too many project directories deployed.');
+        }
+
+        // Save the name of the deployed project folder
+        projectName = contents.shift();
+        debug("Project name:", projectName);
+        // Save the path to this project
+        var projectPath = path.join(dirPath, projectName);
+        // Save a path to where the compiled binary will be
+        var binaryPath = path.join(projectPath, 'target/tessel2/release/');
+
+        // Read the contents of the Cargo.toml to extract binary name
+        // Print out the binary name
+        try {
+          var cargoToml = toml.parse(fs.readFileSync(path.join(projectPath, 'Cargo.toml'), 'utf8'));
+        }
+        catch (error) {
+          debug("Problem parsing Cargo.toml:", error);
+          return rejectRequest(responseCodes.HTTP_BAD_REQUEST, error);
+        }
+
+        // Save the name of the final binary so we can exclude all other files
+        var binaryName = cargoToml.package.name;
```

Do you want to throw an error here if binaryName ends up undefined?